packages/ketcher-core/src/application/editor/actions/erase.ts

```diff
@@ -84,23 +84,30 @@ export function fromOneBondDeletion(restruct, id) {
   return action
 }
 
-export function fromFragmentDeletion(restruct, selection) {
-  assert(!!selection != null)
+export function fromFragmentDeletion(restruct, rawSelection) {
+  assert(!!rawSelection != null)
 
   let action = new Action()
   const atomsToRemove: Array<number> = []
   const frids: Array<number> = []
 
-  selection = {
-    // TODO: refactor me
-    atoms: selection.atoms || [],
-    bonds: selection.bonds || [],
-    rxnPluses: selection.rxnPluses || [],
-    rxnArrows: selection.rxnArrows || [],
-    sgroupData: selection.sgroupData || [],
-    simpleObjects: selection.simpleObjects || [],
-    texts: selection.texts || []
-  }
+  const selection = formatSelection(rawSelection)
+
+  selection.sgroups.forEach((sgroupId) => {
+    const sgroup = restruct.sgroups.get(sgroupId)
+    const sgroupAtoms = sgroup.item.atoms
+
+    selection.atoms = selection.atoms.concat(sgroupAtoms) as any[]
+
+    Object.values(restruct.molecule.bonds).forEach((bond: any, bid) => {
```

Comment:
Looks like you are using iteration index here for `bid`? This may not be a correct bond id.
Also, struct.bonds is a Pool object, are you sure passing it to Object.values() produces a correct result?

Reply:
You're right. Fixed it